server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/crawlers/SeagateCrawler.java

```diff
@@ -0,0 +1,201 @@
+    }
+
+    /**
+     * Process the page, namely get information about parts.
+     * @param page  The fetched page
+     */
+    @Override
+    public void visit(Page page) {
+
+        String url = page.getWebURL().getURL();
+
+        if (page.getParseData() instanceof HtmlParseData) {
+            HtmlParseData htmlParseData = (HtmlParseData) page.getParseData();
+
+            String html = htmlParseData.getHtml();
+            MetadataParser hp = new MetadataParser(html, url);
+            extractData(hp);
+        }
+    }
+
+    /* Extracts metadata from the page. Add found parts to the internal list
+     * of parts.
+     */
+    private void extractData(MetadataParser hp) {
+
+        Elements productNodes = hp.getHtmlDoc().getElementsByClass("node");
+        for (Element productNode : productNodes) {
+
+            // get the product code
+            String productId = hp.selectSingleText(productNode, "div.product-name");
+            if (productId.isEmpty()) {
+                continue;
+            }
+            productId = productId.toUpperCase();
+
+            // get the link to the product, deal with seagate's link problems
+            String link = hp.getLink(productNode);
+            link = fixLink(link);
+
+            // get the image of the product
+            String image = hp.selectSingleAttr(productNode, "img[src]", "abs:src");
+            if (!image.isEmpty()) {
+                image = image.toLowerCase();
+            }
+
+            // get the metadata of the product
+            String description =
+                hp.selectSingleText(productNode, "div.product-description");
+            Size size = extractSize(description);
+
+            if (parts != null) {
+                Part p = new Part(MANUFACTURER_ID, productId, link, image, size,
+                    description);
+                parts.add(p);
+            }
+        }
+    }
+
+    /**
+     * regex hack to deal with seagate's bad links
+     * assumption: all seagate's pdf links follow the format
+     * http://seagateplastics.com/stock_plastics_catalog/
+     * images_catalog/XYZ pdf (1).pdf
+     * where XYZ is the product id. Some of their own (!) urls on their
+     * website are wrong and don't include the " pdf (1)" at the
+     * end ... This is a temporary solution.
+     */
+    private String fixLink(String link) {
+
+        if (FAULTY_PDF_FILTER.matcher(link).matches()) {
+            return link.replaceFirst(".pdf", " pdf (1).pdf");
+        }
+        return link;
+    }
+
+    /**
+     * Given the description on the Seagate website, use standard regex
+     * patterns to attempt to extract the numbers and units.
+     * @return  Size of the part. If none could be parsed, a size with 0x0
+     *          dimensions and unknown unit.
+     */
+    private Size extractSize(String description) {
+
+        for (Pattern p : METADATA_FILTERS) {
+
+            Matcher m = p.matcher(description);
+            if (m.find()) {
```

Comment:
m is dependent on p. It tries a pattern, if that pattern fails, it goes on to the next, and so on. When it finds a match, it breaks the loop, so there's no need for while.